# **Exploratory Data Analisys: Companies**

In the notebook dedicated to the exploratory data analysis (EDA) of companies listed on B3, the filters are specifically designed to meet the needs of investors with a Buy and Hold strategy. These filters allow for better visualization and interpretation of data, facilitating the identification of companies that offer good long-term appreciation and stability, which is essential for this type of investment.

Here are the fundamental metrics that are used to filter and assess the companies:

- **`market_cap`**: to filter companies by their market capitalization, allowing the identification of companies with greater market stability and solidity.
- **`dividend_rate`**: to select companies based on the dividend rate paid, which can be an indicator of consistent cash generation.
- **`trailing_annual_dividend_yield`**: for companies based on past annual dividend yields, a metric that reflects profitability over the stock price.
- **`debt_to_equity`**: to analyze the ratio between the company's total debt and its equity, which helps to understand the level of indebtedness and financial risk.
- **`trailing_pe`**: (Price to Earnings) to compare the current stock price with the earnings per share over the last 12 months, a classic metric for evaluating the price-earnings ratio.
- **`forward_pe`**: for projecting the stock price relative to expected earnings, providing a future estimate of the price-earnings ratio.
- **`price_to_book`**: for comparison between the market value of the company and its book value, which can signal stocks traded below their net asset value.
- **`earnings_quarterly_growth`**: to assess the company's earnings growth compared to the same quarter of the previous year, showing growth or decline trends.

These columns would be at the core of Buy and Hold analysis, as they reflect the fundamental aspects long-term investors generally seek: sustainable growth, financial solidity, and dividend return. The 
visualization of data filtered by these columns offers clear and precise insights, making the investment analysis process more efficient and well-founded.

## **Initial Setup**

### Install Packages

In [ ]:
%pip install pandas -q
%pip install plotly -q

### Import libs

In [1]:
import os
import itertools
import pandas as pd
from pathlib import Path
import plotly.express as px
import plotly.graph_objects as go
import plotly.subplots as sp

### Pandas Config

In [2]:
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

### Create a file path default

In [3]:
file_path_book = str(Path(os.getcwd()).parent.parent.parent / "data/book")

## Companies

### Load data

In [4]:
df_fundamentals_book = pd.read_csv(file_path_book + "/fundamentals_book.csv")
df_fundamentals_book.head(5)

ticker                                          long_name  \
0  ABCB4.SA                              Banco ABC Brasil S.A.   
1  AGRO3.SA  BrasilAgro - Companhia Brasileira de Proprieda...   
2  RAIL3.SA                                          Rumo S.A.   
3  ALPA3.SA                                    Alpargatas S.A.   
4  ALPA4.SA                                    Alpargatas S.A.   

               sector                industry    market_cap  enterprise_value  \
0  Financial Services        Banks - Regional  4.265434e+09      1.477339e+10   
1  Consumer Defensive           Farm Products  2.466480e+09      2.912933e+09   
2         Industrials               Railroads  4.228882e+10      5.524305e+10   
3   Consumer Cyclical  Footwear & Accessories  5.309793e+09      6.482982e+09   
4   Consumer Cyclical  Footwear & Accessories  5.350758e+09      6.395236e+09   

   total_revenue  profit_margins  operating_margins  dividend_rate   beta  \
0   1.941779e+09         0.41576            0.38826           1.56  0.679   
1   1.249437e+09         0.21493            0.25031           3.21  0.432   
2   1.031746e+10         0.07639            0.33544           0.07  0.227   
3   4.022153e+09        -0.05671           -0.06434           0.40  0.571   
4   4.022153e+09        -0.05671           -0.06434           0.43  0.571   

         ebitda  trailing_pe  forward_pe     volume  average_volume  \
0  0.000000e+00     4.069768    4.706601    92300.0        747165.0   
1  2.648920e+08     9.450382    6.332481   298100.0        666692.0   
2  4.522541e+09    54.309525   21.723810  5733400.0      14644522.0   
3 -1.980000e+05     0.000000    0.000000     1100.0          3953.0   
4 -1.980000e+05     0.000000   14.555555  1132100.0       5605825.0   

   fifty_two_week_low  fifty_two_week_high  price_to_sales_trailing_12_months  \
0               15.85                21.99                           2.196663   
1               22.29                32.71                           1.974073   
2               16.21                24.44                           4.098764   
3                7.27                17.80                           1.320137   
4                6.81                22.51                           1.330322   

   fifty_day_average  two_hundred_day_average  trailing_annual_dividend_rate  \
0            19.3382                 18.14667                          1.550   
1            27.0106                 25.58635                          3.240   
2            22.5852                 20.95235                          0.066   
3             8.7146                  9.63540                          0.000   
4             8.3228                  9.27290                          0.000   

   trailing_annual_dividend_yield  book_value  price_to_book    total_cash  \
0                        0.080687      24.518       0.785138  7.774306e+09   
1                        0.132029      22.237       1.113460  3.838370e+08   
2                        0.002993       8.334       2.736981  7.656040e+09   
3                        0.000000       7.867       1.008008  4.142880e+08   
4                        0.000000       7.867       0.999110  4.142880e+08   

   total_cash_per_share    total_debt  earnings_quarterly_growth  \
0                35.162  1.829846e+10                      0.001   
1                 3.885  8.720750e+08                      6.801   
2                 4.132  2.184320e+10                      3.935   
3                 0.614  1.550341e+09                      0.000   
4                 0.614  1.550341e+09                      0.000   

   revenue_growth  gross_margins  ebitda_margins  return_on_assets  \
0           0.003        0.00000         0.00000           0.01530   
1           0.671        0.25252         0.21201           0.03839   
2           0.121        0.34493         0.43834           0.04252   
3          -0.127        0.43246        -0.00005          -0.00910   
4          -0.127        0.43246   

In [5]:
df_fundamentals_numeric_cols = df_fundamentals_book.select_dtypes(include=["int", "number", "float64"])
df_fundamentals_numeric_cols.head(5)

market_cap  enterprise_value  total_revenue  profit_margins  \
0  4.265434e+09      1.477339e+10   1.941779e+09         0.41576   
1  2.466480e+09      2.912933e+09   1.249437e+09         0.21493   
2  4.228882e+10      5.524305e+10   1.031746e+10         0.07639   
3  5.309793e+09      6.482982e+09   4.022153e+09        -0.05671   
4  5.350758e+09      6.395236e+09   4.022153e+09        -0.05671   

   operating_margins  dividend_rate   beta        ebitda  trailing_pe  \
0            0.38826           1.56  0.679  0.000000e+00     4.069768   
1            0.25031           3.21  0.432  2.648920e+08     9.450382   
2            0.33544           0.07  0.227  4.522541e+09    54.309525   
3           -0.06434           0.40  0.571 -1.980000e+05     0.000000   
4           -0.06434           0.43  0.571 -1.980000e+05     0.000000   

   forward_pe     volume  average_volume  fifty_two_week_low  \
0    4.706601    92300.0        747165.0               15.85   
1    6.332481   298100.0        666692.0               22.29   
2   21.723810  5733400.0      14644522.0               16.21   
3    0.000000     1100.0          3953.0                7.27   
4   14.555555  1132100.0       5605825.0                6.81   

   fifty_two_week_high  price_to_sales_trailing_12_months  fifty_day_average  \
0                21.99                           2.196663            19.3382   
1                32.71                           1.974073            27.0106   
2                24.44                           4.098764            22.5852   
3                17.80                           1.320137             8.7146   
4                22.51                           1.330322             8.3228   

   two_hundred_day_average  trailing_annual_dividend_rate  \
0                 18.14667                          1.550   
1                 25.58635                          3.240   
2                 20.95235                          0.066   
3                  9.63540                          0.000   
4                  9.27290                          0.000   

   trailing_annual_dividend_yield  book_value  price_to_book    total_cash  \
0                        0.080687      24.518       0.785138  7.774306e+09   
1                        0.132029      22.237       1.113460  3.838370e+08   
2                        0.002993       8.334       2.736981  7.656040e+09   
3                        0.000000       7.867       1.008008  4.142880e+08   
4                        0.000000       7.867       0.999110  4.142880e+08   

   total_cash_per_share    total_debt  earnings_quarterly_growth  \
0                35.162  1.829846e+10                      0.001   
1                 3.885  8.720750e+08                      6.801   
2                 4.132  2.184320e+10                      3.935   
3                 0.614  1.550341e+09                      0.000   
4                 0.614  1.550341e+09                      0.000   

   revenue_growth  gross_margins  ebitda_margins  return_on_assets  \
0           0.003        0.00000         0.00000           0.01530   
1           0.671        0.25252         0.21201           0.03839   
2           0.121        0.34493         0.43834           0.04252   
3          -0.127        0.43246        -0.00005          -0.00910   
4          -0.127        0.43246        -0.00005          -0.00910   

   return_on_equity  gross_profits  total_assets_approx  asset_turnover  \
0           0.15680   1.973086e+09         7.774306e+09        0.249769   
1           0.12170   3.155040e+08         3.838370e+08        3.255124   
2           0.05163   3.146360e+09         7.656040e+09        1.347623   
3          -0.04153   1.968303e+09         4.142880e+08        9.708591   
4          -0.04153   1.968303e+09         4.142880e+08        9.708591   

   earnings_growth_rate  dividend_payout_ratio        equity  debt_to_equity  \
0                   0.1          155000.000000 -1.052416e+10       -1.738710   
1             

In [6]:
df_top_companies = df_fundamentals_book.nlargest(18, ['market_cap', 'dividend_rate', 'trailing_annual_dividend_yield', 'trailing_annual_dividend_rate', 'debt_to_equity', 'return_on_equity', 'trailing_pe', 'forward_pe', 'price_to_book', 'earnings_quarterly_growth']).drop_duplicates(subset='long_name', keep='first').reset_index(drop=True)
df_top_companies.head(5)

ticker                             long_name              sector  \
0  PETR3.SA  Petróleo Brasileiro S.A. - Petrobras              Energy   
1  VALE3.SA                             Vale S.A.     Basic Materials   
2  ITUB4.SA            Itaú Unibanco Holding S.A.  Financial Services   
3  ABEV3.SA                            Ambev S.A.  Consumer Defensive   
4  BPAC5.SA                Banco BTG Pactual S.A.  Financial Services   

                           industry    market_cap  enterprise_value  \
0              Oil & Gas Integrated  4.832119e+11      7.293640e+11   
1  Other Industrial Metals & Mining  2.776947e+11      3.399071e+11   
2                  Banks - Regional  2.486832e+11      7.077875e+11   
3               Beverages - Brewers  2.007233e+11      1.888459e+11   
4                   Capital Markets  1.483266e+11      4.345411e+10   

   total_revenue  profit_margins  operating_margins  dividend_rate   beta  \
0   5.815630e+11         0.26889            0.39067           9.28  1.097   
1   2.064140e+11         0.27585            0.33672           3.84  0.821   
2   1.189380e+11         0.26661            0.35896           1.50  0.499   
3   8.271054e+10         0.17263            0.18666           0.76  0.604   
4   2.775135e+10         0.30292            0.34960           0.35  1.291   

         ebitda  trailing_pe  forward_pe      volume  average_volume  \
0  2.910870e+11     2.914199    4.821608   2993300.0      13234123.0   
1  8.536400e+10     5.319043    5.421849   4387500.0      25978987.0   
2  0.000000e+00     8.557632    6.867500  10480300.0      24528203.0   
3  2.225154e+10    14.166667   12.750000   8618800.0      24680549.0   
4  0.000000e+00    11.260274    0.000000      2600.0          6184.0   

   fifty_two_week_low  fifty_two_week_high  price_to_sales_trailing_12_months  \
0               23.61                41.86                           0.830885   
1               61.00                98.29                           1.345329   
2               22.62                31.29                           2.090864   
3               12.33                16.88                           2.426816   
4                4.55                 9.08                           5.344842   

   fifty_day_average  two_hundred_day_average  trailing_annual_dividend_rate  \
0            37.0388                 31.98075                          9.657   
1            66.0796                 73.07495                          4.037   
2            27.2770                 26.66895                          0.503   
3            13.4434                 14.10370                          0.762   
4             8.3650                  6.87690                          0.000   

   trailing_annual_dividend_yield  book_value  price_to_book    total_cash  \
0                        0.247298      28.417       1.350600  6.098500e+10   
1                        0.063147      43.051       1.498688  2.423800e+10   
2                        0.018602      18.250       1.505205  3.943250e+11   
3                        0.061254       5.403       2.359800  1.243052e+10   
4                        0.000000       9.808       0.838091  2.211434e+11   

   total_cash_per_share    total_debt  earnings_quarterly_growth  \
0                 4.675  2.793750e+11                     -0.470   
1                 5.580  7.934600e+10                     -0.848   
2                40.237  8.270340e+11                      0.181   
3                 0.789  4.096231e+09                     -0.157   
4                46.436  2.207205e+11                      0.181   

   revenue_growth  gross_margins  ebitda_margins  return_on_assets  \
0          -0.334        0.50633         0.50053           0.15427   
1          -0.131        0.39091         0.41356           0.09975   
2           0.135        0.00000         0.00000           0.01419   
3           0.051        0.49945         0.26903           0.08271   
4           0.021        0.00000         0.0000

In [7]:
df_top_companies_cols_num = df_top_companies.select_dtypes(include=["int", "number", "float64"])
df_top_companies_cols_num.head(5)

market_cap  enterprise_value  total_revenue  profit_margins  \
0  4.832119e+11      7.293640e+11   5.815630e+11         0.26889   
1  2.776947e+11      3.399071e+11   2.064140e+11         0.27585   
2  2.486832e+11      7.077875e+11   1.189380e+11         0.26661   
3  2.007233e+11      1.888459e+11   8.271054e+10         0.17263   
4  1.483266e+11      4.345411e+10   2.775135e+10         0.30292   

   operating_margins  dividend_rate   beta        ebitda  trailing_pe  \
0            0.39067           9.28  1.097  2.910870e+11     2.914199   
1            0.33672           3.84  0.821  8.536400e+10     5.319043   
2            0.35896           1.50  0.499  0.000000e+00     8.557632   
3            0.18666           0.76  0.604  2.225154e+10    14.166667   
4            0.34960           0.35  1.291  0.000000e+00    11.260274   

   forward_pe      volume  average_volume  fifty_two_week_low  \
0    4.821608   2993300.0      13234123.0               23.61   
1    5.421849   4387500.0      25978987.0               61.00   
2    6.867500  10480300.0      24528203.0               22.62   
3   12.750000   8618800.0      24680549.0               12.33   
4    0.000000      2600.0          6184.0                4.55   

   fifty_two_week_high  price_to_sales_trailing_12_months  fifty_day_average  \
0                41.86                           0.830885            37.0388   
1                98.29                           1.345329            66.0796   
2                31.29                           2.090864            27.2770   
3                16.88                           2.426816            13.4434   
4                 9.08                           5.344842             8.3650   

   two_hundred_day_average  trailing_annual_dividend_rate  \
0                 31.98075                          9.657   
1                 73.07495                          4.037   
2                 26.66895                          0.503   
3                 14.10370                          0.762   
4                  6.87690                          0.000   

   trailing_annual_dividend_yield  book_value  price_to_book    total_cash  \
0                        0.247298      28.417       1.350600  6.098500e+10   
1                        0.063147      43.051       1.498688  2.423800e+10   
2                        0.018602      18.250       1.505205  3.943250e+11   
3                        0.061254       5.403       2.359800  1.243052e+10   
4                        0.000000       9.808       0.838091  2.211434e+11   

   total_cash_per_share    total_debt  earnings_quarterly_growth  \
0                 4.675  2.793750e+11                     -0.470   
1                 5.580  7.934600e+10                     -0.848   
2                40.237  8.270340e+11                      0.181   
3                 0.789  4.096231e+09                     -0.157   
4                46.436  2.207205e+11                      0.181   

   revenue_growth  gross_margins  ebitda_margins  return_on_assets  \
0          -0.334        0.50633         0.50053           0.15427   
1          -0.131        0.39091         0.41356           0.09975   
2           0.135        0.00000         0.00000           0.01419   
3           0.051        0.49945         0.26903           0.08271   
4           0.021        0.00000         0.00000           0.01910   

   return_on_equity  gross_profits  total_assets_approx  asset_turnover  \
0           0.39988   3.341000e+11         6.098500e+10        9.536164   
1           0.29588   1.023130e+11         2.423800e+10        8.516132   
2           0.18295   1.155700e+11         3.943250e+11        0.301624   
3           0.16919   3.928676e+10         1.243052e+10        6.653829   
4           0.18284   2.663930e+10         2.211434e+11        0.125490   

   earnings_growth_rate  dividend_payout_ratio        equity  debt_to_equity  \
0                 -47.0               0.000000 -2.183900e+11       -1.279248   
1       

To promote effective portfolio diversification, we refine the dataset to include only one ticker per company, choosing the 10 companies with the best financial metrics.

### Companies Analisys

#### Outlier Analysis

In [ ]:
top_companies_cols = df_top_companies_cols_num.columns
num_columns = len(top_companies_cols)
num_rows = (num_columns + 2) // 3

subplot_titles = [str(col) for col in top_companies_cols]

fig = sp.make_subplots(rows=num_rows, cols=3, subplot_titles=subplot_titles)

for i, column in enumerate(top_companies_cols, start=1):
    row = (i - 1) // 3 + 1
    col = (i - 1) % 3 + 1

    trace = go.Box(y=df_top_companies[column], name=column, 
                   marker_color='lightseagreen', boxpoints='outliers', 
                   jitter=0.7, hoverinfo='y+text', 
                   text=(df_top_companies['ticker']+ ' - ' + df_top_companies['long_name']))

    fig.add_trace(trace, row=row, col=col)

fig.update_layout(title_text='Boxplot of Numerical Variables', height=300*num_rows, showlegend=False, template='plotly_dark')
fig.show()

**Petróleo Brasileiro S.A. - Petrobras: (PETR3.SA)**

- **`market_cap`**:  Petrobras has an atypical market capitalization of **R$ 483,211,900,000**, reflecting its **significant size** in the energy sector.
- **`total_revenue`**: **R$ 581.563 billion**. This metric reflects the total amount of revenue generated by Petrobras in its financial operations.
- **`dividend_rate`**: The dividend rate of Petrobras at **9.28%** of the current stock value is **relatively high**, which may be seen as **positive** for investors seeking to receive **significant dividends**.
- **`ebitda`**: Petrobras recorded an impressive EBITDA of **R$ 291 billion**, highlighting its **solid operational** and **financial performance** in the energy sector.
- **`trailing_annual_dividend_rate`**: The unusual value of **9.65%** for Petrobras may be considered an **outlier**, indicating a **significantly high dividend rate** compared to other companies.
- **`trailing_annual_dividend_yield`**: Petrobras's dividend yield is **0.24%**, showing the **dividends paid to investors** as a percentage of the stock price.
- **`gross_profits`**: The field is **R$ 334.1 billion**, representing the company's total revenue before the deduction of operating expenses, taxes, and other charges.
- **`roce`**: Return on Capital Employed is **0.38171**, indicating the company's efficiency in generating a return on invested capital. The higher the value, the better the efficiency in using its financial resources.

**Itaúsa S.A. (ITSA4.SA)**

- **`profit_margins`**:The field for Itaúsa represents the company's profit margin, which is **1.69976%**. This metric indicates the percentage of profit earned by Itaúsa in relation to its total revenue. That margin could be considered relatively low.
- **`price_to_sales_trailing_12_months`**: The value of 11.09017 for Itaúsa indicates that the market is willing to pay **11.09x**  the company's sales revenue over the past 12 months for each share, which may reflect optimism regarding the company's financial performance.

**Banco BTG Pactual S.A. (BPAC5.SA)**

- **`beta`**:The beta value of 1.291 for Banco BTG Pactual S.A. indicates a relatively high level of volatility in its shares in relation to the market.
- **`debt_to_equity`**: Banco BTG has a debt-to-equity ratio of **521.86**, reflecting a high-leverage strategy with intensive use of debt compared to its equity. This means that for every real of net equity, the bank has approximately **R$ 521,86 in debt**.

**Telefônica Brasil S.A. (VIVT3.SA)**

- **`beta`**: The company Vivo has a beta of **0.061**, indicating that its shares have very low volatility relative to the market, suggesting a less risky investment profile.

**Vale S.A. (VALE3.SA)**

- **`ebitda`**: Vale recorded an EBITDA of **R$ 85.364 billion**, highlighting its solid operational and financial performance in the banking sector.
- **`fifty_two_week_low`**: Vale stock hits 52-week low. Closing value was **R$ 61.00**.

**Centrais Elétricas Brasileiras S.A. - Eletrobrás (ELET3.SA)**

- **`trailing_pe`**: Eletronbrás has a P/E ratio of **34.7**, which is considered relatively high. This means that investors are willing to pay a relatively high price for the company's shares.

**WEG S.A. (WEGE3.SA)**

- **`forward_pe`**: Investors are willing to pay **24.44x times** the company's expected earnings.
_ **`price_to_book`**: Investors are willing to pay a relatively high price of **9x times** the company's book value.

**Banco do Brasil S.A. (BBAS3.SA)**

- **`dividend_payout_ratio`**: The dividend payout ratio (DPR) of **5.17%** is an unusually high number. This means that the company is distributing a very large portion of its profits to shareholders in the form of dividends.

### Histogram

In [ ]:
numeric_cols = df_top_companies_cols_num.columns
subplot_titles = [str(col) for col in numeric_cols]
columns_per_row = 3

num_rows = len(numeric_cols) // columns_per_row + (len(numeric_cols) % columns_per_row > 0)
fig = sp.make_subplots(rows=num_rows, cols=columns_per_row, subplot_titles=subplot_titles)

for i, column in enumerate(numeric_cols):
    row = i // columns_per_row + 1
    col = i % columns_per_row + 1
    
    fig.add_trace(go.Histogram(x=df_top_companies_cols_num[column], name=column, marker_color='lightseagreen'), row=row, col=col)

fig.update_layout(title='Histograms by Column', height=300 * num_rows, showlegend=False, template='plotly_dark')
fig.show()

- **`market_cap`**: Market capitalization ranges from approximately **R$ 74 billion** to **R$ 483 billion** Brazilian reais, with some extremely high values. The market capitalization distribution is positively skewed, with a long right tail in the histogram.
- **`enterprise_value`**: Enterprise value ranges from about **R$ 34 billion** to **R$ 957 billion** Brazilian reais, with significantly different values. The long tail occurs because the exponential distribution is more likely to produce extreme values that are higher than the mean.
- **`total_revenue`**: Total revenue ranges from about **R$ 7.8 billion** to **R$ 581.6 billion** Brazilian reais, with a wide distribution of values. The graph shown has different mean and median, suggesting a lognormal distribution is likely.
- **`profit_margins`**: Profit margins vary from approximately **4.8%** to **27.9%**, showing diversity in margins.
- **`operating_margins`**: Operating margins vary from approximately **0.2%** to **0.46%**, with most values concentrated in the lower range.
- **`dividend_rate`**: Dividend rate varies from **0.22%** to **9.28%**, with most values concentrated around **0.2%** to **0.3%**.
- **`beta`**: Beta values range from about **0.06** to **2.91**, with most values below **1**, suggesting lower volatility compared to the market.
- **`ebitda:`**: EBITDA ranges from approximately **R$ 0** to **R$ 291 billion** Brazilian reais, with highly dispersed values.
- **`trailing_pe`**: Trailing P/E varies from approximately **3.8** to **34.7**, with most values below **10**.
- **`forward_pe`**: Forward P/E varies from approximately **0** to **73.1**, with dispersed values.
- **`volume:`**: Trading volume varies significantly, with some very high values. The distribution is positively skewed.
- **`average_volume`**: Average trading volume ranges from approximately **4.2 million** to **13.2 million**, with most values concentrated in the lower range.
- **`fifty_two_week_low`**: Fifty-two-week lows range from **0** to approximately **66** Brazilian reais.
- **`fifty_two_week_high`**: Fifty-two-week highs range from **0** to approximately **73** Brazilian reais.
- **`price_to_sales_trailing_12_months`**: Price-to-sales ratio for the trailing 12 months ranges from approximately **0.83** to **4.36**, with dispersed values.
- **`fifty_day_average`**: Fifty-day average ranges from approximately **8.36** to **66.08**.
- **`two_hundred_day_average`**: Two-hundred-day average ranges from approximately **6.88** to **73.07**.
- **`trailing_annual_dividend_rate`**: Trailing annual dividend rate ranges from **0** to **9.66** Brazilian reais.
- **`trailing_annual_dividend_yield`**: Trailing annual dividend yield ranges from **0.06%** to **2.47%**.
- **`book_value:`**: Book value ranges from approximately **0.15** to **28.42** Brazilian reais.
- **`price_to_book`**: Price-to-book ratio ranges from approximately **0.39** to **37.04**.
- **`total_cash`**: Total cash ranges from approximately **R$ 4.6 billion** to **R$ 309.8 billion** Brazilian reais.
- **`total_cash_per_share`**: Total cash per share ranges from approximately **0.02** to **46.44** Brazilian reais.
- **`total_debt`**: Total debt ranges from approximately **R$ 1.3 billion** to **R$ 957 billion** Brazilian reais.
- **`earnings_quarterly_growth`**: Quarterly earnings growth ranges from **-84.8%** to **47.2%**.
- **`revenue_growth`**: Revenue growth ranges from **-0.27%** to **0.5%**.
- **`gross_margins`**: Gross margins vary from **0.13%** to **4.5%**.
- **`ebitda_margins`**: EBITDA margins vary from **0.08%** to **0.91%**.
- **`return_on_assets`**: Return on assets varies from **0.08%** to **0.68%**.
- **`return_on_equity`**: Return on equity varies from **0.17%** to **0.68%**.
- **`gross_profits`**: Gross profits vary from approximately **R$ 3.5 billion** to **R$ 581.6 billion** Brazilian reais.
- **`total_assets_approx`**: Total assets approximate range from approximately **R$ 4.1 billion** to **R$ 30.6 billion** Brazilian reais.
- **`asset_turnover`**: Asset turnover varies from **0.02** to **0.08**.
- **`earnings_growth_rate`**: Earnings growth rate varies from **-47%** to **277.9%**.
- **`dividend_payout_ratio`**: Dividend payout ratio varies from **0%** to **134.3%**.
- **`equity`**: Equity varies from approximately **R$ 2.1 billion** to **R$ 30.6 billion** Brazilian reais.
- **`debt_to_equity`**: Debt-to-equity ratio varies from **0.01** to **2.87**.
- **`roi`**: ROI varies from **0.01** to **2.66**.
- **`roce`**: ROCE varies from **-2.35** to **0.80**.


### Correlation and Dispersion

In [ ]:
df_top_companies_corr = df_top_companies_cols_num.corr()
color_scale = [[0, 'rgb(150, 245, 231)'], [0.5, 'rgb(100, 195, 181)'], [1, 'rgb(50, 145, 131)']]
fig = px.imshow(df_top_companies_corr, text_auto=True, aspect="auto", width=3000, height=1000, template="plotly_dark", title="Mapa de Calor: Correlação entre as variaveis", color_continuous_scale=color_scale)
fig.show()

**Correlações de Crescimento dos Lucros**:

- A correlação perfeita **(1.000000)** entre **`earnings_growth_rate`** e **`earnings_quarterly_growth`** significa que qualquer variação nos lucros de um trimestre para o outro é refletida igualmente na taxa anual de crescimento dos lucros. Isso pode ser útil para prever tendências futuras de lucratividade baseadas em dados históricos de crescimento.

**Correlações de Médias Móveis de Ações**:

- A alta correlação entre **`two_hundred_day_average`**, **`fifty_day_average`**, e os extremos do mercado (`fifty_two_week_low` e `fifty_two_week_high`) com valores acima de **0.980** indica que as médias móveis de ações são **indicadores confiáveis** para identificar **tendências de longo prazo** e podem ser usadas para sinalizar pontos de **entrada** e **saída** de investimentos.

**Correlações de Rendimento Corporativo**:

- **`total_revenue`** tem uma **correlação forte** com **`ebitda`** e **`gross_profits`**, ilustrando que empresas com **maiores receitas** geralmente relatam **maiores lucros brutos e EBITDA**, que é um indicador de eficiência operacional e capacidade de **geração de lucro**.

**Dinâmica de Valor de Mercado**:

- A correlação entre **`market_cap`** e variáveis como **`total_revenue`**, gross_profits, e dividend_rate (variando de 0.870635 a 0.943719) mostra como o valor de mercado de uma empresa é influenciado por sua performance financeira e política de dividendos. Isso sugere que investidores valorizam a capacidade de uma empresa de sustentar crescimento de receita e lucros, assim como recompensar acionistas com dividendos.

**Relações de Dividendos e Desempenho**:

- A correlação entre trailing_annual_dividend_rate, trailing_annual_dividend_yield, e dividend_rate indica que as empresas que oferecem maiores dividendos tendem a ter um rendimento anual acumulado mais alto, o que pode atrair investidores que buscam renda regular de dividendos.

**Indicadores de Eficiência e Retorno**:

- roce está correlacionado com dividend_rate e ebitda, sugerindo que o retorno sobre o capital empregado é uma métrica importante que pode impactar tanto a capacidade de uma empresa de pagar dividendos quanto sua eficiência operacional geral.

**Correlações Negativas de Patrimônio e Dívida**:

- Uma forte correlação negativa entre equity e total_debt (-0.941708) indica que empresas com alto patrimônio líquido tendem a ter menos dívida. Isso pode ser interpretado como um sinal de solidez financeira e menor alavancagem, o que é frequentemente visto como um ponto positivo pelos investidores.

**Correlações Negativas de Liquidez e Margens**:

- total_cash_per_share demonstra uma correlação negativa com gross_margins e ebitda_margins, destacando que uma liquidez por ação maior não necessariamente se traduz em margens de lucro maiores. Isso pode ocorrer em cenários onde a liquidez é mantida como uma reserva estratégica ou usada em investimentos que não geram retorno imediato.

**Inter-relações de Dívida, Ativos e Rentabilidade**:

- total_debt está fortemente relacionado com total_assets_approx (0.722928), o que pode indicar que empresas maiores com mais ativos tendem a ter mais dívida, o que é normal dada a necessidade de financiamento para expansão e crescimento.

**Correlações Diversas**:

- Outras correlações notáveis incluem a relação entre roi e roce, e return_on_assets e return_on_equity, sugerindo que diferentes métricas de retorno estão alinhadas e que melhorias em uma área de desempenho financeiro podem levar a melhorias em outras.

### Análise de Valor de Mercado e Valor da Empresa

In [ ]:
df_top_companies.sort_values(by = 'market_cap', ascending=True, inplace=True)

fig = px.bar(df_top_companies, title='Market Cap', x='long_name', y='market_cap', color_discrete_sequence=['rgb(100, 195, 181)'], template='plotly_dark', hover_name='ticker', height=600)
formatted_market_cap = [f'R${x:.2f}B' for x in (df_top_companies['market_cap'] / 1e9)]
fig.update_traces(text=formatted_market_cap, textposition='outside', textfont=dict(color='white'))
fig.update_xaxes(title='Companies')
fig.update_yaxes(title='Market Cap (R$)')
fig.show()

In [ ]:
df_top_companies.sort_values(by = 'enterprise_value', ascending=True, inplace=True)

fig = px.bar(df_top_companies, title='Enterprise Value', x='long_name', y='enterprise_value', color_discrete_sequence=['rgb(100, 195, 181)'], template='plotly_dark', hover_name='ticker', height=600)
formatted_enterprise_value = [f'R${x:.2f}B' for x in (df_top_companies['enterprise_value'] / 1e9)]
fig.update_traces(text=formatted_enterprise_value, textposition='outside', textfont=dict(color='white'))
fig.update_xaxes(title='Companies')
fig.update_yaxes(title='Enterprise Value (R$)')
fig.show()

### Rentabilidade e Margens:

In [ ]:
df_top_companies.sort_values(by = 'total_revenue', ascending=True, inplace=True)

fig = px.bar(df_top_companies, title='Total Revenue', x='long_name', y='total_revenue', color_discrete_sequence=['rgb(100, 195, 181)'], template='plotly_dark', hover_name='ticker', height=600)
formatted_total_revenue = [f'R${x:.2f}B' for x in (df_top_companies['total_revenue'] / 1e9)]
fig.update_traces(text=formatted_total_revenue, textposition='outside', textfont=dict(color='white'))
fig.update_xaxes(title='Companies')
fig.update_yaxes(title='Total Revenue (R$)')
fig.show()

In [ ]:
df_top_companies.sort_values(by = 'profit_margins', ascending=True, inplace=True)

fig_op_margin = px.bar(df_top_companies, title='Profit Margin', x='long_name', y='profit_margins', color_discrete_sequence=['rgb(100, 195, 181)'], hover_name=df_top_companies.ticker, height=600)
fig_op_margin.update_traces(text=[f'{x:.2f}%' for x in (df_top_companies['profit_margins']*100)], textposition='outside',textfont=dict(color='white'))
fig_op_margin.update_layout(title='Profit Margin', xaxis_title='Sector', yaxis_title='Profit Margins (%)', template = 'plotly_dark')
fig_op_margin.show()

In [ ]:
df_top_companies.sort_values(by = 'operating_margins', ascending=True, inplace=True)

fig_op_margin = px.bar(df_top_companies, title='Operating Margin', x='long_name', y='operating_margins', color_discrete_sequence=['rgb(100, 195, 181)'], hover_name=df_top_companies.ticker, height=600)
fig_op_margin.update_traces(text=[f'{x:.2f}%' for x in (df_top_companies['operating_margins']*100)], textposition='outside',textfont=dict(color='white'))
fig_op_margin.update_layout(xaxis_title='Sector', yaxis_title='Profit Margins (%)', template = 'plotly_dark')
fig_op_margin.show()

### Avaliação de Desempenho e Crescimento:

In [ ]:
df_top_companies.sort_values(by = 'trailing_pe', ascending=True, inplace=True)

fig_op_margin = px.bar(df_top_companies, title='Trailing PE', x='long_name', y='trailing_pe', color_discrete_sequence=['rgb(100, 195, 181)'], hover_name=df_top_companies.ticker, height=600)
fig_op_margin.update_traces(text=[f'{x:.2f}' for x in (df_top_companies['trailing_pe'])], textposition='outside',textfont=dict(color='white'))
fig_op_margin.update_layout(xaxis_title='Sector', yaxis_title='Trailing P/E', template = 'plotly_dark')
fig_op_margin.show()

In [ ]:
df_top_companies.sort_values(by = 'forward_pe', ascending=True, inplace=True)

fig_op_margin = px.bar(df_top_companies, title='Forward P/E', x='long_name', y='forward_pe', color_discrete_sequence=['rgb(100, 195, 181)'], hover_name=df_top_companies.ticker, height=600)
fig_op_margin.update_traces(text=[f'{x:.2f}%' for x in (df_top_companies['forward_pe'])], textposition='outside',textfont=dict(color='white'))
fig_op_margin.update_layout(xaxis_title='Companies', yaxis_title='Forward P/E', template = 'plotly_dark')
fig_op_margin.show()

In [ ]:
df_top_companies.sort_values(by = 'earnings_quarterly_growth', ascending=True, inplace=True)

fig_op_margin = px.bar(df_top_companies, title='Earnings Quarterly Growth', x='long_name', y='earnings_quarterly_growth', color_discrete_sequence=['rgb(100, 195, 181)'], hover_name=df_top_companies.ticker, height=600)
fig_op_margin.update_traces(text=[f'{x:.2f}%' for x in (df_top_companies['earnings_quarterly_growth']*10)], textposition='outside',textfont=dict(color='white'))
fig_op_margin.update_layout(xaxis_title='Companies', yaxis_title='Earnings Quarterly Growth (%)', template = 'plotly_dark')
fig_op_margin.show()

In [ ]:
df_top_companies.sort_values(by = 'revenue_growth', ascending=True, inplace=True)

fig_op_margin = px.bar(df_top_companies, title='Revenue Growth', x='long_name', y='revenue_growth', color_discrete_sequence=['rgb(100, 195, 181)'], hover_name=df_top_companies.ticker, height=600)
fig_op_margin.update_traces(text=[f'{x:.2f}%' for x in (df_top_companies['revenue_growth']*100)], textposition='outside',textfont=dict(color='white'))
fig_op_margin.update_layout(xaxis_title='Companies', yaxis_title='Revenue Growth (%)', template = 'plotly_dark')
fig_op_margin.show()

### Eficiência e Uso de Capital:

In [ ]:
df_top_companies.sort_values(by = 'return_on_assets', ascending=True, inplace=True)

fig_op_margin = px.bar(df_top_companies, title='Return on Assets', x='long_name', y='return_on_assets', color_discrete_sequence=['rgb(100, 195, 181)'], hover_name=df_top_companies.ticker, height=600)
fig_op_margin.update_traces(text=[f'{x:.2f}%' for x in (df_top_companies['return_on_assets']*100)], textposition='outside',textfont=dict(color='white'))
fig_op_margin.update_layout(xaxis_title='Companies', yaxis_title='Return on Assets (%)', template = 'plotly_dark')
fig_op_margin.show()

In [ ]:
df_top_companies.sort_values(by = 'return_on_equity', ascending=True, inplace=True)

fig_op_margin = px.bar(df_top_companies, title='Return on Equity', x='long_name', y='return_on_equity', color_discrete_sequence=['rgb(100, 195, 181)'], hover_name=df_top_companies.ticker, height=600)
fig_op_margin.update_traces(text=[f'{x:.2f}%' for x in (df_top_companies['return_on_equity']*100)], textposition='outside',textfont=dict(color='white'))
fig_op_margin.update_layout(xaxis_title='Companies', yaxis_title='Return on Equity (%)', template = 'plotly_dark')
fig_op_margin.show()

In [ ]:
df_top_companies.sort_values(by = 'debt_to_equity', ascending=True, inplace=True)

fig_op_margin = px.bar(df_top_companies, title='Debt to Equity', x='long_name', y='debt_to_equity', color_discrete_sequence=['rgb(100, 195, 181)'], hover_name=df_top_companies.ticker, height=600)
fig_op_margin.update_traces(text=[f'{x:.2f}%' for x in (df_top_companies['debt_to_equity']*10)], textposition='outside',textfont=dict(color='white'))
fig_op_margin.update_layout(xaxis_title='Companies', yaxis_title='Debt to Equity (%)', template = 'plotly_dark')
fig_op_margin.show()

In [8]:
df_top_companies.sort_values(by = 'roi', ascending=True, inplace=True)

fig_op_margin = px.bar(df_top_companies, title='ROI', x='long_name', y='roi', color_discrete_sequence=['rgb(100, 195, 181)'], hover_name=df_top_companies.ticker, height=600)
fig_op_margin.update_traces(text=[f'{x:.2f}%' for x in (df_top_companies['roi']*10)], textposition='outside',textfont=dict(color='white'))
fig_op_margin.update_layout(xaxis_title='Companies', yaxis_title='ROI (%)', template = 'plotly_dark')
fig_op_margin.show()

In [9]:
df_top_companies.sort_values(by = 'roce', ascending=True, inplace=True)

fig_op_margin = px.bar(df_top_companies, title='ROCE', x='long_name', y='roce', color_discrete_sequence=['rgb(100, 195, 181)'], hover_name=df_top_companies.ticker, height=600)
fig_op_margin.update_traces(text=[f'{x:.2f}%' for x in (df_top_companies['roce']*10)], textposition='outside',textfont=dict(color='white'))
fig_op_margin.update_layout(xaxis_title='Companies', yaxis_title='ROCE (%)', template = 'plotly_dark')
fig_op_margin.show()

### Liquidez e Saúde Financeira:

In [17]:
df_top_companies.sort_values(by = 'total_cash', ascending=True, inplace=True)

fig_op_margin = px.bar(df_top_companies, title='Total Cash', x='long_name', y='total_cash', color_discrete_sequence=['rgb(100, 195, 181)'], hover_name=df_top_companies.ticker, height=600)
fig_op_margin.update_traces(text=[f'R${x:.2f}B' for x in (df_top_companies['total_cash']/1000000000)], textposition='outside',textfont=dict(color='white'))
fig_op_margin.update_layout(xaxis_title='Companies', yaxis_title='Total Cash', template = 'plotly_dark')
fig_op_margin.show()

In [18]:
df_top_companies.sort_values(by = 'total_debt', ascending=True, inplace=True)

fig_op_margin = px.bar(df_top_companies, title='Total Debt', x='long_name', y='total_debt', color_discrete_sequence=['rgb(100, 195, 181)'], hover_name=df_top_companies.ticker, height=600)
fig_op_margin.update_traces(text=[f'R${x:.2f}B' for x in (df_top_companies['total_debt']/1000000000)], textposition='outside',textfont=dict(color='white'))
fig_op_margin.update_layout(xaxis_title='Companies', yaxis_title='Total Debt', template = 'plotly_dark')
fig_op_margin.show()

### Valuation de Ações

In [21]:
df_top_companies.sort_values(by = 'price_to_book', ascending=True, inplace=True)

fig_op_margin = px.bar(df_top_companies, title='Total Debt', x='long_name', y='price_to_book', color_discrete_sequence=['rgb(100, 195, 181)'], hover_name=df_top_companies.ticker, height=600)
fig_op_margin.update_traces(text=[f'{x:.2f}x' for x in (df_top_companies['price_to_book'])], textposition='outside',textfont=dict(color='white'))
fig_op_margin.update_layout(xaxis_title='Companies', yaxis_title='Price To Book', template = 'plotly_dark')
fig_op_margin.show()

In [24]:
df_top_companies.sort_values(by = 'price_to_sales_trailing_12_months', ascending=True, inplace=True)

fig_op_margin = px.bar(df_top_companies, title='Total Debt', x='long_name', y='price_to_sales_trailing_12_months', color_discrete_sequence=['rgb(100, 195, 181)'], hover_name=df_top_companies.ticker, height=600)
fig_op_margin.update_traces(text=[f'{x:.2f}x' for x in (df_top_companies['price_to_sales_trailing_12_months'])], textposition='outside',textfont=dict(color='white'))
fig_op_margin.update_layout(xaxis_title='Companies', yaxis_title='Price To Sale', template = 'plotly_dark')
fig_op_margin.show()

### Dividendos:

In [25]:
df_top_companies.sort_values(by = 'dividend_rate', ascending=True, inplace=True)

fig_op_margin = px.bar(df_top_companies, title='Dividend Rate', x='long_name', y='dividend_rate', color_discrete_sequence=['rgb(100, 195, 181)'], hover_name=df_top_companies.ticker, height=600)
fig_op_margin.update_traces(text=[f'{x:.2f}x' for x in (df_top_companies['dividend_rate'])], textposition='outside',textfont=dict(color='white'))
fig_op_margin.update_layout(xaxis_title='Companies', yaxis_title='Dividend Rate', template = 'plotly_dark')
fig_op_margin.show()

In [37]:
df_top_companies.sort_values(by = 'dividend_payout_ratio', ascending=True, inplace=True)

fig_op_margin = px.bar(df_top_companies, title='Dividend Payout Ratio', x='long_name', y='dividend_payout_ratio', color_discrete_sequence=['rgb(100, 195, 181)'], hover_name=df_top_companies.ticker, height=600)
fig_op_margin.update_traces(text=[f'{x:.2f}%' for x in (df_top_companies['dividend_payout_ratio']/10)], textposition='outside',textfont=dict(color='white'))
fig_op_margin.update_layout(xaxis_title='Companies', yaxis_title='Dividend Rate', template = 'plotly_dark')
fig_op_margin.show()

### Análise de Tendências de Mercado:

In [42]:
df_top_companies.sort_values(by = 'beta', ascending=True, inplace=True)

fig_op_margin = px.bar(df_top_companies, title='beta', x='long_name', y='beta', color_discrete_sequence=['rgb(100, 195, 181)'], hover_name=df_top_companies.ticker, height=600)
fig_op_margin.update_traces(text=[f'{x:.2f}' for x in (df_top_companies['beta'])], textposition='outside',textfont=dict(color='white'))
fig_op_margin.update_layout(xaxis_title='Companies', yaxis_title='Beta', template = 'plotly_dark')
fig_op_margin.show()

In [52]:
df_top_companies.sort_values(by = 'average_volume', ascending=True, inplace=True)

fig_op_margin = px.bar(df_top_companies, title='Volume', x='long_name', y='average_volume', color_discrete_sequence=['rgb(100, 195, 181)'], hover_name=df_top_companies.ticker, height=600)
fig_op_margin.update_traces(text=[f'{x:.2f}M' for x in (df_top_companies['average_volume']/1000000)], textposition='outside',textfont=dict(color='white'))
fig_op_margin.update_layout(xaxis_title='Companies', yaxis_title='average_volume', template = 'plotly_dark')
fig_op_margin.show()

### Comparação de Preço de Ações

In [53]:
df_top_companies

ticker                                         long_name  \
4   BPAC5.SA                            Banco BTG Pactual S.A.   
10  VIVT3.SA                            Telefônica Brasil S.A.   
7   WEGE3.SA                                          WEG S.A.   
6   BBAS3.SA                              Banco do Brasil S.A.   
9   ELET3.SA  Centrais Elétricas Brasileiras S.A. - Eletrobrás   
0   PETR3.SA              Petróleo Brasileiro S.A. - Petrobras   
8   ITSA4.SA                                       Itaúsa S.A.   
2   ITUB4.SA                        Itaú Unibanco Holding S.A.   
3   ABEV3.SA                                        Ambev S.A.   
1   VALE3.SA                                         Vale S.A.   
5   BBDC4.SA                               Banco Bradesco S.A.   

                    sector                          industry    market_cap  \
4       Financial Services                   Capital Markets  1.483266e+11   
10  Communication Services                  Telecom Services  7.645024e+10   
7              Industrials    Specialty Industrial Machinery  1.384426e+11   
6       Financial Services                  Banks - Regional  1.404060e+11   
9                Utilities             Utilities - Renewable  7.929263e+10   
0                   Energy              Oil & Gas Integrated  4.832119e+11   
8              Industrials                     Conglomerates  8.658098e+10   
2       Financial Services                  Banks - Regional  2.486832e+11   
3       Consumer Defensive               Beverages - Brewers  2.007233e+11   
1          Basic Materials  Other Industrial Metals & Mining  2.776947e+11   
5       Financial Services                  Banks - Regional  1.423388e+11   

    enterprise_value  total_revenue  profit_margins  operating_margins  \
4       4.345411e+10   2.775135e+10         0.30292            0.34960   
10      8.845781e+10   5.031185e+10         0.08991            0.16545   
7       1.303664e+11   3.175831e+10         0.15827            0.20575   
6       9.574406e+11   9.532893e+10         0.34830            0.45908   
9       1.201184e+11   3.551173e+10         0.04797            0.42603   
0       7.293640e+11   5.815630e+11         0.26889            0.39067   
8       9.397528e+10   7.807000e+09         1.69976            0.20113   
2       7.077875e+11   1.189380e+11         0.26661            0.35896   
3       1.888459e+11   8.271054e+10         0.17263            0.18666   
1       3.399071e+11   2.064140e+11         0.27585            0.33672   
5       5.421110e+11   7.289319e+10         0.20718            0.22324   

    dividend_rate   beta        ebitda  trailing_pe  forward_pe      volume  \
4            0.35  1.291  0.000000e+00    11.260274    0.000000      2600.0   
10           2.58  0.061  1.510794e+10    17.029520   13.613569    518500.0   
7            0.32  0.513  6.480928e+09    26.829268   24.444445   8227600.0   
6            0.94  0.803  0.000000e+00     3.770115    3.805105   1529200.0   
9            0.22  0.629  1.274860e+10    34.700000    8.361445    867000.0   
0            9.28  1.097  2.910870e+11     2.914199    4.821608   2993300.0   
8            0.74  0.647  2.240000e+09     6.740458    5.553459   2919900.0   
2            1.50  0.499  0.000000e+00     8.557632    6.867500  10480300.0   
3            0.76  0.604  2.225154e+10    14.166667   12.750000   8618800.0   
1            3.84  0.821  8.536400e+10     5.319043    5.421849   4387500.0   
5            0.23  0.675  0.000000e+00    10.649254    6.150862  14450200.0   

    average_volume  fifty_two_week_low  fifty_two_week_high  \
4           6184.0                4.55                 9.08   
10       2252017.0               35.63                46.83   
7        7064393.0               31.47                42.42   
6        8738341.0               30.25                51.99   
9        8876606.0               29.81                52.49   
0       13234123.0               23.61                41.86 

In [54]:
df_top_companies.sort_values(by = 'fifty_two_week_low', ascending=True, inplace=True)

fig_op_margin = px.bar(df_top_companies, title='fifty_two_week_low', x='long_name', y='fifty_two_week_low', color_discrete_sequence=['rgb(100, 195, 181)'], hover_name=df_top_companies.ticker, height=600)
fig_op_margin.update_traces(text=[f'{x:.2f}M' for x in (df_top_companies['fifty_two_week_low'])], textposition='outside',textfont=dict(color='white'))
fig_op_margin.update_layout(xaxis_title='Companies', yaxis_title='fifty_two_week_low', template = 'plotly_dark')
fig_op_margin.show()

In [55]:
df_top_companies.sort_values(by = 'fifty_two_week_high', ascending=True, inplace=True)

fig_op_margin = px.bar(df_top_companies, title='fifty_two_week_high', x='long_name', y='fifty_two_week_high', color_discrete_sequence=['rgb(100, 195, 181)'], hover_name=df_top_companies.ticker, height=600)
fig_op_margin.update_traces(text=[f'{x:.2f}M' for x in (df_top_companies['fifty_two_week_high'])], textposition='outside',textfont=dict(color='white'))
fig_op_margin.update_layout(xaxis_title='Companies', yaxis_title='fifty_two_week_low', template = 'plotly_dark')
fig_op_margin.show()

In [56]:

df_top_companies.sort_values(by = 'fifty_day_average', ascending=True, inplace=True)

fig_op_margin = px.bar(df_top_companies, title='fifty_day_average', x='long_name', y='fifty_day_average', color_discrete_sequence=['rgb(100, 195, 181)'], hover_name=df_top_companies.ticker, height=600)
fig_op_margin.update_traces(text=[f'{x:.2f}M' for x in (df_top_companies['fifty_day_average'])], textposition='outside',textfont=dict(color='white'))
fig_op_margin.update_layout(xaxis_title='Companies', yaxis_title='fifty_day_average', template = 'plotly_dark')
fig_op_margin.show()

In [57]:


df_top_companies.sort_values(by = 'two_hundred_day_average', ascending=True, inplace=True)

fig_op_margin = px.bar(df_top_companies, title='two_hundred_day_average', x='long_name', y='two_hundred_day_average', color_discrete_sequence=['rgb(100, 195, 181)'], hover_name=df_top_companies.ticker, height=600)
fig_op_margin.update_traces(text=[f'{x:.2f}M' for x in (df_top_companies['two_hundred_day_average'])], textposition='outside',textfont=dict(color='white'))
fig_op_margin.update_layout(xaxis_title='Companies', yaxis_title='two_hundred_day_average', template = 'plotly_dark')
fig_op_margin.show()

## TL / DR